core/lib/zksync_core/src/state_keeper/io/mod.rs

```diff
@@ -96,209 +136,107 @@ pub trait StateKeeperIO: 'static + Send + IoSealCriteria {
-    /// Marks the L1 batch as sealed.
-    async fn seal_l1_batch(
+    /// Loads base system contracts with the specified version.
+    async fn load_base_system_contracts(
         &mut self,
-        witness_block_state: Option<WitnessBlockState>,
-        updates_manager: UpdatesManager,
-        l1_batch_env: &L1BatchEnv,
-        finished_batch: FinishedL1Batch,
-    ) -> anyhow::Result<()>;
-
-    /// Loads protocol version of the previous l1 batch.
-    async fn load_previous_batch_version_id(&mut self) -> anyhow::Result<ProtocolVersionId>;
+        protocol_version: ProtocolVersionId,
+        cursor: &IoCursor,
+    ) -> anyhow::Result<BaseSystemContracts>;
+    /// Loads protocol version of the specified L1 batch, which is guaranteed to exist in the storage.
+    async fn load_batch_version_id(
+        &mut self,
+        number: L1BatchNumber,
+    ) -> anyhow::Result<ProtocolVersionId>;
     /// Loads protocol upgrade tx for given version.
     async fn load_upgrade_tx(
         &mut self,
         version_id: ProtocolVersionId,
     ) -> anyhow::Result<Option<ProtocolUpgradeTx>>;
 }
 
-impl fmt::Debug for dyn StateKeeperIO {
-    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
-        formatter
-            .debug_struct("StateKeeperIO")
-            .field("current_l1_batch_number", &self.current_l1_batch_number())
-            .field("current_miniblock_number", &self.current_miniblock_number())
-            .finish()
+/// Handler for state keeper outputs (miniblocks and L1 batches).
+#[async_trait]
+pub trait HandleStateKeeperOutput: 'static + Send + fmt::Debug {
+    /// Initializes this handler. This method will be called on state keeper initialization before any other calls.
+    /// The default implementation does nothing.
+    async fn initialize(&mut self, _cursor: &IoCursor) -> anyhow::Result<()> {
+        Ok(())
     }
-}
 
-/// A command together with the return address allowing to track command processing completion.
-#[derive(Debug)]
-struct Completable<T> {
-    command: T,
-    completion_sender: oneshot::Sender<()>,
+    /// Handles a miniblock (aka L2 block) produced by the state keeper.
+    async fn handle_miniblock(&mut self, updates_manager: &UpdatesManager) -> anyhow::Result<()>;
+
+    /// Handles an L1 batch produced by the state keeper.
+    async fn handle_l1_batch(
+        &mut self,
+        _witness_block_state: Option<&WitnessBlockState>,
+        _updates_manager: &UpdatesManager,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
 }
 
-/// Handle for [`MiniblockSealer`] allowing to submit [`MiniblockSealCommand`]s.
 #[derive(Debug)]
-pub struct MiniblockSealerHandle {
-    commands_sender: mpsc::Sender<Completable<MiniblockSealCommand>>,
-    latest_completion_receiver: Option<oneshot::Receiver<()>>,
-    // If true, `submit()` will wait for the operation to complete.
-    is_sync: bool,
+pub struct CompoundOutputHandler {
```

Comment:
Nit: I like the concept, but I'm not sure if it's a good thing that you have two ways to achieve the same thing (e.g. by providing `Box<dyn HandleStateKeeperOutput>` directly or creating `CompoundOutputHandler` with no `with_aux_handler` invocation.

What if we will:
1. Rename this structure to simply `OutputHandler`.
2. Make `ZkSyncStateKeeper` accept `OutputHandler` instead of `Box<dyn HandleStateKeeperOutput>`

This way this structure will become a single _primary_ concept. Moreover, in the docs for it we'll be able to cover *both* the concept of `HandleStateKeeperOutput` and composability, without making people to find out about these separately.

(at the same time, internally, the state keeper can still have `Box<dyn HandleStateKeeperOutput>`)

Reply:
I've had similar thoughts, will do.